tools/run_tests/helper_scripts/clean_xds_interop_resources.py

```diff
@@ -0,0 +1,170 @@
+        proc.kill()
+        outs, errs = proc.communicate()
+    if proc.returncode:
+        logging.error('> Failed to execute cmd [%s], returned %d, stderr: %s',
+                      " ".join(cmds), proc.returncode, errs)
+        return None
+    return outs
+
+
+def exec_gcloud(*cmds: List[str]) -> Json:
+    cmds = [GCLOUD, '--project', PROJECT] + list(cmds)
+    if 'list' in cmds:
+        # Add arguments to shape the list output
+        cmds.extend([
+            '--format', 'json', '--filter',
+            f'creationTimestamp<={get_expire_timestamp()}'
+        ])
+    # Executing the gcloud command
+    return json.loads(exec_command(*cmds))
+
+
+def remove_relative_resources_run_xds_tests(suffix: str):
+    """Removing GCP resources created by run_xds_tests.py."""
+    logging.info('Removing run_xds_tests.py resources with suffix [%s]', suffix)
+    cmds = [
+        sys.executable, RUN_XDS_TESTS, '--clean_only',
+        f'--gcp_suffix="{suffix}"', '--verbose'
+    ]
+    if args.dry_run:
+        logging.debug('> Skipped: %s', " ".join(cmds))
+        return
+    exec_command(cmds, cwd=ROOT)
+
+
+def remove_relative_resources_psm_sec(prefix: str):
+    """Removing GCP resources created by PSM Sec framework."""
+    logging.info('Removing PSM Security resources with prefix [%s]', prefix)
+    cmds = [
+        sys.executable,
+        '-m',
+        'bin.run_td_setup',
+        '--cmd=cleanup',
+        '--flagfile=config/grpc-testing.cfg',
+        f'--namespace={prefix}',
+        # Following arguments doesn't matter
+        '--kube_context=DUMMY_DOES_MATTER',
+        '--server_image=DUMMY_DOES_MATTER',
+        '--client_image=DUMMY_DOES_MATTER'
+    ]
+    if args.dry_run:
+        logging.debug('> Skipped: %s', " ".join(cmds))
+        return
+    exec_command(cmds,
+                 cwd=os.path.join(ROOT, 'tools', 'run_tests',
+                                  'xds_k8s_test_driver'))
+
+
+def check_one_type_of_gcp_resources(list_cmd: List[str],
+                                    old_framework_residual_search: str = '',
+                                    new_framework_residual_search: str = ''):
+    logging.info('Checking GCP resources with %s or %s',
+                 old_framework_residual_search, new_framework_residual_search)
+    for resource in exec_gcloud(*list_cmd):
+        if old_framework_residual_search:
+            result = re.search(old_framework_residual_search, resource['name'])
+            if result is not None:
+                remove_relative_resources_run_xds_tests(result.group(1))
+                continue
+
+        if new_framework_residual_search:
+            result = re.search(new_framework_residual_search, resource['name'])
+            if result is not None:
+                remove_relative_resources_psm_sec(result.group(1))
+                continue
+
+
+def check_costly_gcp_resources() -> None:
+    check_one_type_of_gcp_resources(
+        ['compute', 'forwarding-rules', 'list'],
```

Comment:
Is there actually any need to check forwarding rules, target-grpc-proxies, and target-http-proxies? Since all of those will require (transitively) a url-map, wouldn't a search for url-maps be enough?

Reply:
Your comment does not address anything about my comment. I was saying that if there is a target-grpc-proxy, then there is also a url-map, so we can just do the url map query. If we find anything, then it will go and delete all the types.

Reply:
Emm... they are different resources. If I try to delete a url map, it will fail if the corresponding target-proxy is still alive. I know the search is redundant in most cases, where the resource creation and deletion is atomic. But in real life:

1. People might hit CTRL-C during the resource creation or deletion;
2. People might try to delete the resources manually, and found it hard to do so, then give up after deleting a few.

Reply:
Since url-map _must_ to be created first and deleted last, how could you have a target-proxy without having a url-map? I don't see how the "real life" examples invalidate my argument.

Reply:
You are right. The url-map can exist without a target-proxy, but not vice versa. So, we can skip the search for target proxies. Updated.

Reply:
I had also mentioned forwarding rules.

Url-maps require a backend service as well, but I'm not sure we want to depend on that since the backend service name formats could change pretty easily.

Reply:
Updated to only look for url-maps, and removed unnecessary functions. Thanks for the suggestion!

(Our backend service quota is much larger than others, 500.) Emm... the naming format won't be changed frequently in both frameworks. Otherwise, their own clean-up won't work as good.

Searching for url-maps can free 10 sets of resources, which is good enough. But if we consider searching for unused backend services, we can release 5 additional backend services.